src/node_contextify.cc

```diff
@@ -1705,14 +1705,32 @@ static void ContainsModuleSyntax(const FunctionCallbackInfo<Value>& args) {
 
   CHECK_GE(args.Length(), 2);
 
-  // Argument 1: source code
-  CHECK(args[0]->IsString());
-  Local<String> code = args[0].As<String>();
-
   // Argument 2: filename
   CHECK(args[1]->IsString());
   Local<String> filename = args[1].As<String>();
 
+  // Argument 1: source code; if undefined, read from filename in argument 2
+  Local<String> code;
+  if (args[0]->IsUndefined()) {
+    CHECK(!filename.IsEmpty());
+    const char* filename_str = Utf8Value(isolate, filename).out();
+    std::string contents;
+    int result = ReadFileSync(&contents, filename_str);
```

Comment:
Reading from disk from the native layer here doesn’t seem right - it’s not guaranteed that this is on disk and at a location pointed to by filename (the filename can be some artificial ID). The JS land should be refactored to only call this after source code is confirmed instead.

Reply:
this comes as part of the trials to cover the module type resolution for the `resolve`. I would expect if any of that happens (correct me if I'm wrong) that `ReadFileSync` will return `!= 0` [here](https://github.com/nodejs/node/blob/main/src/util.cc#L275) and this would cause the early exit with an exception.

Additionally we call this from managed code [here](https://github.com/dygabo/node/blob/fix-for-unflagging-module-format-detection/lib/internal/modules/esm/get_format.js#L91-L93). I am guessing that at this stage the fileUrl was already validated but that is at this stage just an assumption. [This test](https://github.com/dygabo/node/blob/fix-for-unflagging-module-format-detection/test/es-module/test-esm-detect-ambiguous.mjs#L428) made me think that.

Reply:
The exception would be surprising for a hook - resolve shouldn't be poking into the file system, that should be load's job.

Reply:
that is changed now, the comment above is obsolete. The function would return undefined on file access errors. But I am still looking for a way to remove this file access without losing the test.